PublicMapping/districtbuilder
# Allow for selecting partially locked districts

## Overview
Allow for selecting partially locked districts.

### Checklist


### Demo
Selecting/deselecting locked counties
![db_partially_locked_final](https://user-images.githubusercontent.com/2926237/93124548-c68f5c00-f697-11ea-8869-4e83ae78a2ab.gif)

Partial selection with locked blocks (messaging will be added to improve the UX of this in https://github.com/PublicMapping/districtbuilder/issues/417)
![db_partially_locked_selection_ui_weirdness](https://user-images.githubusercontent.com/2926237/93124660-ee7ebf80-f697-11ea-923b-d1c6cb1a1902.gif)

### Notes
There were a couple of existing bugs that I ran into which had to be smoothed out to get everything working:
* For the rectangle selection tool, `getFeaturesInBoundingBox` only returned features at the currently selected geolevel, though since we made the change to allow for working with multiple levels at once, we need to get all features from all levels (that's what the new `getAllSelectedFeatures` does)
* `isGeoUnitLocked` wasn't quite working as advertised and would quit recursing too early if the district definition was jagged which could result in incorrect behavior

Much of the meat of the changes was to `onlyUnlockedGeoUnits` which needed to be modified to break out a geounit such as a county into its constituent unlocked sub-geounits. I kept things imperative-style since the `Map` API is imperative, for better or for worse. I'm not totally sure about the name `removeLockedGeoUnits` since it removes geounits that are partially locked and _adds_ the unlocked child geounits, but I think it does get the idea across and there is a docstring to try to explain it.

There is a good amount of complexity within the selection tools. There already was a good amount of complexity to the rectangle selection tool, though it didn't have to change all that much (complexity stayed about the same). The default selection tool did need to become significantly more complex because it needed to support the selection _and_ deselection of partially locked geounits. I added quite a few comments there to make that as clear as I could.

I ended up adding a helper function `getChildGeoUnits` which is used liberally for implementing this functionality. It returns child geounits but no further (no grandchildren). This seems fine for now since we're only dealing with breaking up counties into blockgroups at the moment. This function is also now used instead of the now defunct `findSelectedSubFeatures` which queried the Mapbox map. Given the limitations around Mapbox only returning features within the current viewport it seems preferable to use our own static metadata for feature querying whenever possible (and I would think it would be faster though I'm not sure about that).

## Testing Instructions
* Select a blockgroup and assign it to a district
* Lock that district
* Using the default selection tool, select the county containing the locked blockgroup
* Ensure all non-locked blockgroups are selected
* Deselect the county and all of the blockgroups should be deselected
* Select a single blockgroup and then select the whole county containing the blockgroup
* Ensure the single blockgroup is not double-selected
* Assign a block to the locked district
* Try to select the blockgroup which contains it (nothing should happen)
* Try to select the county which contains the blockgroup (all the blockgroups except the one containing the locked block should be selected)

Closes #323
